File: src/main/java/me/taubsie/carrylogs/application/service/ServerService.java
package me.taubsie.carrylogs.application.service;

import me.taubsie.carrylogs.application.classes.ServerData;
import me.taubsie.carrylogs.application.classes.ServerProperty;
import me.taubsie.carrylogs.application.start.BotStarter;
import me.taubsie.dungeonhub.common.CarryLogService;
import me.taubsie.dungeonhub.common.OnStart;
import me.taubsie.dungeonhub.common.ProgramOrigin;
import me.taubsie.dungeonhub.common.StartupListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;
import java.io.File;
import java.io.IOException;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Time;
import java.util.*;

@OnStart
public class ServerService implements StartupListener {
    private static final Logger logger = LoggerFactory.getLogger(ServerService.class);
    private static ServerService instance;
    private final Set<ServerData> serverData = new HashSet<>();
    private Timer timer;

    public static ServerService getInstance() {
        if(instance == null) {
            instance = new ServerService();
        }

        return instance;
    }

    private ServerService() {
        try {
            Files.createDirectory(Path.of(getServerFolder()));
        } catch(FileAlreadyExistsException ignored) {
            //Ignored since I just want to be sure that the folder always exists.
        } catch(IOException ioException) {
            ioException.printStackTrace();
        }
    }

    private void loadServers() {
        serverData.clear();

        BotStarter.getInstance()
                .getBot()
                .getServers()
                .forEach(server -> loadServerData(server.getId()));
    }

    private void resetTimer() {
        if(timer != null) {
            timer.cancel();
        }

        timer = new Timer();

        timer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                logger.debug("Server configs reloaded!");
                loadServers();
            }
        }, new Time(System.currentTimeMillis() + 1000 * 60 * 15), 1000L * 60 * 15);
    }

    public String getServerFolder() {
        return CarryLogService.getInstance().getMainFolder() + File.separator + "servers";
    }

    public void loadServerData(long id) {
        serverData.add(new ServerData(id));
    }

    public void unloadServerData(long id) {
        serverData.removeIf(serverData1 -> serverData1.getId() == id);
    }

    public Optional<ServerData> getServerData(long id) {
        return serverData.stream()
                .filter(data -> data.getId() == id)
                .findAny();
    }

    public String getServerProperty(long id, ServerProperty serverProperty) {
        return getServerData(id)
                .map(data -> data.getConfig(serverProperty))
                .filter(s -> !s.isBlank())
                .orElse(serverProperty.getDefaultValue());
    }

    public Optional<String> getActualServerProperty(long id, ServerProperty serverProperty) {
        return getServerData(id)
                .map(data -> data.getConfig(serverProperty))
                .filter(s -> !s.isBlank());
    }

    public boolean canUse(long id, @Nullable ServerProperty serverProperty) {
        //TODO finish implementation (if needed here)
        if(serverProperty == null) {
            return false;
        }

        return getServerData(id)
                .map(serverData1 -> serverData1.isEnabled(serverProperty))
                .orElse(false);
    }

    @Override
    public void onStart(ProgramOrigin programOrigin) {
        loadServers();

        resetTimer();
    }
}

File: src/main/java/me/taubsie/dungeonhub/kord/application/service/ServerService.kt
package me.taubsie.dungeonhub.kord.application.service

import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import me.taubsie.dungeonhub.kord.application.misc.ServerData
import me.taubsie.dungeonhub.common.DungeonHubService
import me.taubsie.dungeonhub.kord.application.connection.DiscordConnection
import me.taubsie.dungeonhub.kord.application.enums.ServerProperty
import me.taubsie.dungeonhub.kord.application.loader.OnStart
import me.taubsie.dungeonhub.kord.application.loader.StartupListener
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.nio.file.FileAlreadyExistsException
import java.nio.file.Files
import java.nio.file.Path
import java.sql.Time
import java.util.*
import java.util.function.Predicate

@OnStart
object ServerService : StartupListener {
    private val logger: Logger = LoggerFactory.getLogger(ServerService::class.java)
    private val serverData: MutableSet<ServerData> = HashSet()
    private var timer: Timer? = null

    init {
        try {
            Files.createDirectory(Path.of(serverFolder))
        } catch (ignored: FileAlreadyExistsException) {
            //Ignored since I just want to be sure that the folder always exists.
        } catch (ioException: IOException) {

            logger.error(null, ioException)
        }
    }

    private suspend fun loadServers() {
        serverData.clear()

        DiscordConnection.bot?.kordRef?.guilds
            ?.collect { server ->
                loadServerData(
                    server.id.value.toLong()
                )
            }
    }

    private fun resetTimer() {
        if (timer != null) {
            timer!!.cancel()
        }

        timer = Timer()

        timer!!.scheduleAtFixedRate(object : TimerTask() {
            override fun run() {
                logger.debug("Server configs reloaded!")
                runBlocking {
                    launch {
                        loadServers()
                    }
                }
            }
        }, Time(System.currentTimeMillis() + 1000 * 60 * 15), 1000L * 60 * 15)
    }

    val serverFolder: String
        get() = DungeonHubService.getInstance().mainFolder + File.separator + "servers"

    fun loadServerData(id: Long) {
        serverData.add(ServerData(id))
    }

    fun unloadServerData(id: Long) {
        serverData.removeIf { serverData1: ServerData -> serverData1.id == id }
    }

    fun getServerData(id: Long): Optional<ServerData> {
        return serverData.stream()
            .filter { data: ServerData -> data.id == id }
            .findAny()
    }

    val allServers: Set<ServerData>
        get() = serverData

    fun getServersWhere(function: Predicate<ServerData>?): List<ServerData> {
        return serverData.stream()
            .filter(function)
            .toList()
    }

    fun getActualServerProperty(id: Long, serverProperty: ServerProperty): Optional<String> {
        return getServerData(id)
            .map { data: ServerData ->
                data.getConfig(
                    serverProperty
                )
            }
            .flatMap { s -> Optional.ofNullable(s) }
            .filter { s -> s.isNotBlank() }
    }

    fun canUse(id: Long, serverProperty: ServerProperty?): Boolean {
        //TODO finish implementation (if needed here)
        if (serverProperty == null) {
            return false
        }

        return getServerData(id)
            .map { serverData1: ServerData ->
                serverData1.isEnabled(
                    serverProperty
                )
            }
            .orElse(false)
    }

    override suspend fun postStart() {
        loadServers()

        resetTimer()
    }
}